File: backend/adapter-web/src/main/java/id/behavio/web/PublicHost.java
package id.behavio.web;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

/**
 * Host yang dipakai untuk MENYEBUT simulator ke dunia luar — contoh {@code curl} di
 * dashboard & {@code servers[].url} di export OpenAPI.
 *
 * <p><b>Kenapa ada.</b> Host-host itu dulu string harfiah {@code "localhost"} di kode
 * (frontend & {@code OpenApiExporter}), jadi contoh curl hanya berguna di mesin yang
 * menjalankan Behavio — padahal gunanya justru untuk ditempel ke Postman/klien di mesin
 * lain. {@code server.forward-headers-strategy} tak menolong sedikit pun: setelan itu
 * hanya berlaku saat Spring merekonstruksi URL dari request, sementara tak ada yang
 * bertanya — hostnya dihardcode.
 *
 * <p><b>Urutan resolusi</b> (yang pertama menang):
 * <ol>
 *   <li>{@code DEPLOY_HOST} / properti {@code behavio.public-host} — deklarasi eksplisit
 *       operator. Wajib menang: hanya operator yang tahu alamat publik saat port
 *       simulator dipetakan berbeda (container, NAT, proxy).</li>
 *   <li>Host dari request yang sedang berjalan. Di sinilah
 *       {@code forward-headers-strategy: framework} akhirnya berguna — Spring membaca
 *       {@code X-Forwarded-Host}, sehingga host yang dipakai = host yang diketik user
 *       di browser (mis. IP LAN mesin ini), bukan loopback server.</li>
 *   <li>{@code localhost} — hanya kalau tak ada request terikat (mis. dipanggil dari
 *       job/test).</li>
 * </ol>
 *
 * <p><b>Yang TIDAK bisa diperbaiki di sini:</b> port. Contoh curl menunjuk <i>port
 * simulator</i> (mis. 9101) yang dilayani {@code HttpServer} JDK tersendiri dan tak
 * pernah melewati Spring — jadi kalau reverse proxy Anda hanya meneruskan :8080, host
 * yang benar saja tak membuat port itu terjangkau. Petakan port simulator di proxy, atau
 * set {@code DEPLOY_HOST} ke alamat yang memang menjangkaunya.
 */
@Component
public class PublicHost {

    private static final String FALLBACK = "localhost";

    private final String configured;

    public PublicHost(@Value("${behavio.public-host:}") String configured) {
        this.configured = configured == null ? "" : configured.trim();
    }

    /** Host publik saat ini — aman dipanggil di luar konteks request. */
    public String resolve() {
        if (!configured.isBlank()) {
            return configured;
        }
        try {
            String host = ServletUriComponentsBuilder.fromCurrentRequest().build().getHost();
            if (host != null && !host.isBlank()) {
                return host;
            }
        } catch (IllegalStateException e) {
            // tak ada request terikat ke thread ini — jatuh ke fallback
        }
        return FALLBACK;
    }
}

File: frontend/src/app/core/api/public-host.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject, signal } from '@angular/core';

/**
 * Host yang dipakai contoh `curl` di dashboard.
 *
 * Dulu setiap contoh curl menulis `http://localhost:<port>` harfiah, jadi ia hanya
 * berguna di mesin yang menjalankan Behavio — padahal gunanya justru untuk ditempel ke
 * Postman/klien di mesin LAIN. Menambahkan `forward-headers-strategy` di Spring tak
 * menolong: setelan itu hanya berlaku kalau ada yang bertanya host ke Spring, sementara
 * string ini dikompilasi ke bundle browser.
 *
 * Urutan: `DEPLOY_HOST` dari backend (satu-satunya yang tahu alamat publik saat port
 * simulator dipetakan lewat NAT/proxy) → host yang sedang dipakai browser → `localhost`.
 * Backend sudah menerapkan urutan yang sama; `location.hostname` di sini hanya jaring
 * pengaman bila /config gagal diambil.
 */
@Injectable({ providedIn: 'root' })
export class PublicHost {
  private readonly http = inject(HttpClient);

  /** Dipakai template curl; terisi ulang begitu /config terjawab. */
  readonly host = signal<string>(PublicHost.browserHost());

  private loaded = false;

  private static browserHost(): string {
    return typeof window !== 'undefined' && window.location.hostname
      ? window.location.hostname
      : 'localhost';
  }

  /** Idempoten — aman dipanggil dari tiap halaman yang butuh. */
  load(): void {
    if (this.loaded) return;
    this.loaded = true;
    this.http.get<{ publicHost?: string }>('/api/admin/v1/config').subscribe({
      next: cfg => {
        if (cfg?.publicHost) this.host.set(cfg.publicHost);
      },
      error: () => {
        /* pertahankan host browser — tetap jauh lebih berguna daripada localhost */
      },
    });
  }
}
